### PR TITLE
Cross-plugin task: Add logging for webhook events #813

### LINKED ISSUE #813: Cross-plugin task: Add logging for webhook events
Same task as https://github.com/mattermost/mattermost-plugin-github/pull/487

### PR DESCRIPTION
**Pull request by sibasankarnayak**
Added custom setting for Jira plugin to enable the logs for webhook event

ticket here
Fixes #813

### CODE REVIEW

File: plugin.json [hunk 1]
```diff
@@ -100,6 +100,13 @@
         "type": "bool",
         "help_text": "Display subscription name in post when a subscription posts to a channel",
         "default": false
+      },
+      {
+        "key": "EnableWebhookEventLogging",
+        "display_name": "Enable Webhook Event Logging:",
```
Comment (mickmister) [inline plugin.json:106]: Seems weird how we have colons for some of the `display_name` but not all. @sibasankarnayak Do you mind adding colons to the rest of the settings as well? Also the periods at the end of the `help_text`

File: server/webhook_http.go [hunk 2]
```diff
@@ -74,6 +75,15 @@ func (p *Plugin) httpWebhook(w http.ResponseWriter, r *http.Request, instanceID
 	if err != nil {
 		return respondErr(w, status, err)
 	}
+
+	if conf.EnableWebhookEventLogging {
+		parsedRequest, eventErr := httputil.DumpRequest(r, true)
+		if eventErr != nil {
+			return respondErr(w, status, eventErr)
+		}
+		p.API.LogDebug("Webhook Event Log", "event", string(parsedRequest))
```
Comment (mickmister) [inline server/webhook_http.go:84]: @sibasankarnayak Can you post an example of the output in a comment on this PR?

**Review by mickmister — CHANGES_REQUESTED** [review]
LGTM, just one request to change from `LogInfo` to `LogDebug`

File: plugin.json [hunk 3]
```diff
@@ -96,10 +96,17 @@
...
+        "display_name": "Display subscription name in notifications:",
         "type": "bool",
         "help_text": "Display subscription name in post when a subscription posts to a channel",
         "default": false
+      },
+      {
+        "key": "EnableWebhookEventLogging",
+        "display_name": "Enable Webhook Event Logging:",
+        "type": "bool",
+        "help_text": "Allow the plugin to log the event from webhook.",
```
Comment (mickmister) [inline plugin.json:108]: Can we also state that the log level needs to be set to `DEBUG` in order to see these webhook events in the logs?

File: server/subscribe.go [hunk 4]
```diff
@@ -714,11 +715,19 @@ func (p *Plugin) httpSubscribeWebhook(w http.ResponseWriter, r *http.Request, in
...
 	if err != nil {
 		return respondErr(w, status, err)
 	}
 
+	if conf.EnableWebhookEventLogging {
+		parsedRequest, eventErr := httputil.DumpRequest(r, true)
+		if eventErr != nil {
+			return respondErr(w, status, eventErr)
+		}
+		p.API.LogInfo("Webhook Event Log", "event", string(parsedRequest))
```
Comment (mickmister) [inline server/subscribe.go:729]: This needs to change to `LogDebug` too

```suggestion
		p.API.LogDebug("Webhook Event Log", "event", string(parsedRequest))
```

**Review by mickmister — CHANGES_REQUESTED** [review]

File: plugin.json [hunk 5]
```diff
@@ -96,10 +96,17 @@
...
+        "display_name": "Display subscription name in notifications:",
         "type": "bool",
         "help_text": "Display subscription name in post when a subscription posts to a channel",
         "default": false
+      },
+      {
+        "key": "EnableWebhookEventLogging",
+        "display_name": "Enable Webhook Event Logging:",
+        "type": "bool",
+        "help_text": "Allow the plugin to log the event from webhook and the log level needs to be set to DEBUG.",
```
Comment (jfrerich) [inline plugin.json:108]:
```suggestion
        "help_text": "Allow the plugin to log the webhook event. The log level needs to be set to DEBUG.",
```

**Review by jfrerich — CHANGES_REQUESTED** [review]
I have a small rewording suggestion

File: plugin.json (hunk 5 above)
Comment (mickmister) [reply plugin.json]: If this supports markdown can we surround `DEBUG` with backticks?

```go
    "help_text": "Allow the plugin to log the webhook event. The log level needs to be set to `DEBUG`.",
```

**Review by mickmister — APPROVED** [review]
LGTM :+1:

File: plugin.json (hunk 5 above)
Comment (sibasankarnayak) [reply plugin.json]: @mickmister sure , it looks great!

image

**Review by jfrerich — APPROVED** [review]

**Comment by sibasankarnayak** [comment]
@dipak-demansol ,  @DHaussermann  can you please review this PR

**Comment by dipak-demansol** [comment]
> @dipak-demansol , @DHaussermann can you please review this PR

i'll make sure it will done max till tomorrow.

**Comment by dipak-demansol** [comment]
@sibasankarnayak @hanzei i tested the PR and its LGTM but After Discussing with Dylan there are more condition with that we want to test and to Complete testing i need the admin access of Jira account so after that i'm able to complete the testing, whenever i got that access, i will Complete the Testing.

**Comment by dipak-demansol** [comment]
i have admin access, i will check it.

**Review by dipak-demansol — APPROVED** [review]
**LGTM** 
# i Tested The Logs with below Condition and **Compared the PR logs with the Master Branch.**   (**Only tested with the Cloud instance)**
   1. when plugin was disable.
   2. When app was not installed in the Jira.
   2.1 Changing the WebHook + status API secret key.

**Pull request merged**